test/test_apply_rhs.jl

```diff
@@ -0,0 +1,75 @@
+function test_apply_rhs()
+    grid = generate_grid(Quadrilateral, (20, 20))
+    dim = 2
+    ip = Lagrange{dim,RefCube,1}()
+    qr = QuadratureRule{dim,RefCube}(2)
+    cellvalues = CellScalarValues(qr, ip)
+    
+    dh = DofHandler(grid)
+    push!(dh, :u, 1)
+    close!(dh)
+    
+    K = create_sparsity_pattern(dh)
+    
+    ch = ConstraintHandler(dh)
+    
+    ∂Ω = union(getfaceset.((grid,), ["left", "right", "top", "bottom"])...)
+    
+    dbc = Dirichlet(:u, ∂Ω, (x, t) -> 0)
+    add!(ch, dbc);
+    
+    close!(ch)
+    update!(ch, 0.0);
+    
+    function doassemble(
+        cellvalues::CellScalarValues{dim},
+        K::SparseMatrixCSC,
+        dh::DofHandler,
+    ) where {dim}
+    
+        n_basefuncs = getnbasefunctions(cellvalues)
+        Ke = zeros(n_basefuncs, n_basefuncs)
+        fe = zeros(n_basefuncs)
+    
+        f = zeros(ndofs(dh))
+        assembler = start_assemble(K, f)
+    
+        @inbounds for cell in CellIterator(dh)
+            fill!(Ke, 0)
+            fill!(fe, 0)
+    
+            reinit!(cellvalues, cell)
+    
+            for q_point = 1:getnquadpoints(cellvalues)
+                dΩ = getdetJdV(cellvalues, q_point)
+    
+                for i = 1:n_basefuncs
+                    v = shape_value(cellvalues, q_point, i)
+                    ∇v = shape_gradient(cellvalues, q_point, i)
+                    fe[i] += v * dΩ
+                    for j = 1:n_basefuncs
+                        ∇u = shape_gradient(cellvalues, q_point, j)
+                        Ke[i, j] += (∇v ⋅ ∇u) * dΩ
+                    end
+                end
+            end
+    
+            assemble!(assembler, celldofs(cell), fe, Ke)
+        end
+        return K, f
+    end
+    
+    K, f = doassemble(cellvalues, K, dh)
+    A = create_sparsity_pattern(dh)
+    A, g = doassemble(cellvalues, A, dh)
+    rhsdata = get_rhs_data(ch, A)
+    
+    apply!(K, f, ch)
+    apply_rhs!(rhsdata, g, ch)
+    u₁ = K \ f
+    u₂ = K \ g
```

Comment:
Considering that this is the only example for usage of `apply_rhs!` so far, I think it would be nice to be clear here. Using `apply_rhs!` requires that the stiffness matrix has been modified once as well.

```suggestion
    apply!(K, f, ch)
    apply!(A, ch) #need to apply bcs to A once
    apply_rhs!(rhsdata, g, ch)
    u₁ = K \ f
    u₂ = A \ g
```

(But maybe `apply(A, ch)` could be made abundant for the user by making it happen in a joint step with generating the rhs_data? What do you think?)

Reply:
yes if people are confused by the interface we can still wrap it in `get_rhs_data`. However, I've encountered this kind of interface in various FE frameworks, so I think it's convenient to do `apply!(A,ch)` before the time loop and `apply_rhs!` right before each solve statement in the time loop